### ISSUE TITLE
Respect core.add-remotes-config-dir configuration option when adding remotes

### ISSUE THREAD

**Issue by pwithnall** [flatpak/flatpak#1665]
There are currently a couple of places in `common/flatpak-dir.c` where `ostree_repo_write_config()` is called to add per-remote configuration to the OSTree configuration. This won’t work if the repository is configured with `core.add-remotes-config-dir=true`. For the moment, OSTree sets `core.add-remotes-config-dir=false` by default for non-system repositories (such as flatpak repositories). However, on systems where the OSTree and flatpak repositories are shared (such as Endless OS), this can cause problems if `core.add-remotes-config-dir=false` isn’t explicitly configured. Problems like `flatpak remote-add` failing:
```
$ sudo flatpak remote-add new-remote http://example.com/
error: Remote "new-remote" already defined in /etc/flatpak/remotes.d/new-remote.conf
```

The problem is that `ostree_repo_remote_[add|change]()` will preferentially write remote config to `/etc/ostree/remotes.d/` (or wherever the configured remote-config-dir is), if `core.add-remotes-config-dir=true`. But `ostree_repo_write_config()` will always modify `$repo_dir/config`.

Since late last year, calls to `ostree_repo_write_config()` will error out if `core.add-remotes-config-dir=true` is set, but the config provided to it contains remotes which are already defined in `/etc/ostree/remotes.d`. This prevents the system getting into a bad state, but it does mean that `flatpak remote-add` is broken (it will not succeed, and will leave a partially-configured remote behind when it fails).

Two functions in `common/flatpak-dir.c` are currently affected:
 * `flatpak_dir_modify_remote()`
 * `flatpak_dir_install_bundle()`

For example, if I have:
```
$ cat /ostree/repo/config && ls -laR /etc/*/remotes.d
[core]
repo_version=1
mode=bare
add-remotes-config-dir=true
min-free-space-percent=0

[remote "flathub"]
url=https://dl.flathub.org/repo/
xa.title=Flathub
gpg-verify=true
gpg-verify-summary=true

/etc/ostree/remotes.d:
total 8
drwxr-xr-x 2 root root 4096 May 11 10:59 .
drwxr-xr-x 3 root root 4096 May 11 10:59 ..
```

Then I run:
```
$ sudo flatpak remote-add new-remote http://example.com/
error: Remote "new-remote" already defined in /etc/flatpak/remotes.d/new-remote.conf
```

I end up with:
```
$ cat /ostree/repo/config && ls -laR /etc/*/remotes.d
[core]
repo_version=1
mode=bare
add-remotes-config-dir=true
min-free-space-percent=0

[remote "flathub"]
url=https://dl.flathub.org/repo/
xa.title=Flathub
gpg-verify=true
gpg-verify-summary=true

/etc/flatpak/remotes.d:
total 12
drwxr-xr-x 2 root root 4096 May 11 14:23 .
drwxr-xr-x 3 root root 4096 May 11 14:18 ..
-rw-r--r-- 1 root root   46 May 11 14:23 new-remote.conf

/etc/ostree/remotes.d:
total 8
drwxr-xr-x 2 root root 4096 May 11 10:59 .
drwxr-xr-x 3 root root 4096 May 11 10:59 ..
$ cat /etc/flatpak/remotes.d/new-remote.conf 
[remote "new-remote"]
url=http://example.com/
```

**Comment by dbnicholson** [comment]
Another option is to finish off https://github.com/ostreedev/ostree/pull/1166 so that flatpak can do sane remote modification instead of doing the "add remote with just URL if it doesn't exist then set options via `ostree_repo_copy_config()/ostree_repo_write_config()` dance".

**Comment by pwithnall** [comment]
That would be the best option. I don’t have time to work on that at the moment though. :-(

**Comment by mwleeds** [comment]
@pwithnall this can be closed, right?

**Comment by pwithnall** [comment]
> @pwithnall this can be closed, right?

Nope, the right fix is to modify the file which contains the particular remote config, rather than unconditionally modifying `$repo_dir/config`. The ‘fix’ I pushed in https://github.com/flatpak/flatpak/commit/21e219bf1dff3258fc7c451a13910939b61264ea just returns an error message in the dodgy case, rather than actually doing the right thing.

The long-term fix here is probably https://github.com/ostreedev/ostree/pull/1166, but I’ll leave it for Alex to decide whether to defer to 1166, or whether to fix this in flatpak in the interim.

**Comment by Mikenux** [comment]
@pwithnall: Can't this be closed now, as it appears the long term fix has been merged?

**Comment by beedell-roke** [comment]
In case it matters, this would remediate [`src.fedoraproject.org/rpms/flatpak/pull-request/23`](https://src.fedoraproject.org/rpms/flatpak/pull-request/23#comment-323914), too.
